### PR TITLE
#213 Add ftp endpoint

### PR DESCRIPTION
**Pull request by pinaf**
Issue #213 
 - added dependency to apache commons-net
 - FtpClientTest: tests commons-net library by connecting to ftp.gnu.org and downloading a README file
 - FtpFacadeTest: connectDisconnect
 - Main
  - puzzle: Create a FtpFacade in order to provide a FTP gateway. Also unignore test 'connectDisconnect' in FtpFacadeTest.
 - FtpFacade
  - puzzle: Implement TLS secure port listening in a manner analogous to HttpFacade.
 - FtpResponseTest: constructsCorrectly
 - FtpResponse
  - puzzle: Implement methods 'withCode', 'withText', 'send' as above. Also, implement 'asString' and unignore test 'constructsCorrectly' in FtpResponseTest.
 - FtpThread
  - puzzle: Implement authentication based on the USER and PASS commands.
  - puzzle: Implement relay functionality to fetch resource upon receival of the RETR command with valid authorization.

### CODE REVIEW

**Comment by davvd** [comment]
Let me find a reviewer for this pull request, thanks for submitting it

**Comment by davvd** [comment]
@krzyk it's yours, please review

**Comment by pinaf** [comment]
@krzyk weird, ftp client test passes here. any way to trigger it again? probably a network issue.

**Comment by pinaf** [comment]
@krzyk I'm new with the puzzle creation process so I'd appreciate any feedback :)

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+ * <p>The class is instantiated in {@link Main}, once per application run.
+ *
+ * <p>The class is immutable and thread-safe.
+ *
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ * @see Main
+ * @checkstyle ClassDataAbstractionCoupling (10 lines)
+ * @todo 213 Implement TLS secure port listening in a manner analogous to
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: todos should have format; `@todo #number:time Description`, so e.g. `@todo #213:30min Some text` or `@todo #213:1hr Other text`.

File: s3auth-relay/pom.xml [hunk 2]
```diff
@@ -174,6 +174,11 @@
             <groupId>com.google.guava</groupId>
             <artifactId>guava</artifactId>
         </dependency>
+        <dependency>
+            <groupId>commons-net</groupId>
+            <artifactId>commons-net</artifactId>
+            <version>3.3</version>
```
Comment (krzyk) [inline s3auth-relay/pom.xml]: if a project inherits from jcabi parent, you don't have to include versions for common libraries (look at `dependencyManagement` section in https://github.com/jcabi/jcabi-parent/blob/master/pom.xml)

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+ *
+ * <p>The class is instantiated in {@link Main}, once per application run.
+ *
+ * <p>The class is immutable and thread-safe.
+ *
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ * @see Main
+ * @checkstyle ClassDataAbstractionCoupling (10 lines)
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: is this required?

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpResponse.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * <p>By default FTP code is OK (200) and content is empty.
+ *
+ * <p>The class is NOT thread-safe.
+ *
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ * @see FtpThread
+ * @todo 213 Implement methods 'withCode', 'withText', 'send' as above.
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpResponse.java]: please correct todo format

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+ * queue, processing it, and closing the socket. The class is instantiated
+ * by {@link FtpFacade} and is executed by Services Executor routinely.
+ *
+ * <p>The class is thread-safe.
+ *
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ * @see FtpFacade
+ * @todo 213 Implement authentication based on the USER and PASS commands.
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: puzzle format correction needed

File: s3auth-relay/src/main/java/com/s3auth/relay/Main.java [hunk 6]
```diff
@@ -64,6 +64,8 @@ private Main() {
      * Entrance.
      * @param args Optional arguments
      * @throws Exception If something is wrong
+     * @todo 213 Create a FtpFacade in order to provide a FTP gateway. Also
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/Main.java]: todo format

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        this.sockets = sckts;
+        this.hosts = hsts;
+    }
+
+    /**
+     * Dispatch one request from the encapsulated queue.
+     * @return Amount of bytes sent to socket
+     * @throws InterruptedException If interrupted while waiting for the queue
+     * @checkstyle ExecutableStatementCount (100 lines)
+     * @checkstyle MultipleStringLiterals (100 lines)*
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: the start at the end can be removed

File: s3auth-relay/src/test/java/com/s3auth/relay/FtpClientTest.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+    /**
+     * FTPClient can connect and run basic commands (USER, LIST, RETR, QUIT).
+     * @throws IOException If there is some problem inside
+     */
+    @Test
+    public void runBasicCommands() throws IOException {
+        final FTPClient ftp = new FTPClient();
+        final String server = "ftp.gnu.org";
+        final String filename = "README";
```
Comment (krzyk) [inline s3auth-relay/src/test/java/com/s3auth/relay/FtpClientTest.java]: you can inline this

File: s3auth-relay/src/test/java/com/s3auth/relay/FtpClientTest.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+import org.junit.Test;
+
+/**
+ * Test case for the apache commons-net FTP client.
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @checkstyle LineLength (150 lines)
+ * @checkstyle ExecutableStatementCount (30 lines)
+ * @version $Id$
+ */
+public final class FtpClientTest {
```
Comment (krzyk) [inline s3auth-relay/src/test/java/com/s3auth/relay/FtpClientTest.java]: You are testing commons-net FtpClient here? I don't see any mention of s3auth classes here (maybe I'm missing something)
If so then please remove this, we should assume that apache commons folks did that already :)

File: s3auth-relay/src/test/java/com/s3auth/relay/FtpFacadeTest.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * Connects and disconnects.
+     * @throws IOException If it fails inside
+     */
+    @Test
+    @Ignore
+    public void connectDisconnect() throws IOException {
+        final FTPClient ftp = new FTPClient();
+        ftp.connect("localhost");
```
Comment (krzyk) [inline s3auth-relay/src/test/java/com/s3auth/relay/FtpFacadeTest.java]: Please add at least setup of `FtpFacade` here, right now it looks like we are connecting to localhost and doing nothing.
Even if the testcase is `@Ignore`d we should (at least to our knowledge, before implementing all the functionality) create a test that should work if someone implements the right todos.

**Comment by krzyk** [comment]
@pinaf Few comments above, and alse one more general. For me it looks like we could create a base class for all protocols and make `HttpFacade` and `FtpFacade` inherit from it. (please remember that you can make a PDD puzzle for this). Do you agree?

File: s3auth-relay/pom.xml (hunk 2 above)
Comment (pinaf) [inline s3auth-relay/pom.xml]: @krzyk I see, but when I take it out I get import errors on the commons-net classes I need. Maybe it's beacause jcabi-parent is a parent of `s3auth` but not of `s3auth-relay`?

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java (hunk 3 above)
Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: Yes it is.
```
[ERROR] /src/main/java/com/s3auth/relay/FtpFacade.java[67]: Class Data Abstraction Coupling is 9 (max allowed is 7) classes [FtpFacade.FTPThreadRunnable, FtpResponse, FtpThread, IOException, Runnable, ServerSocket, SynchronousQueue, VerboseRunnable, VerboseThreads]. (ClassDataAbstractionCouplingCheck)
```

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java (hunk 1 above)
Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: Got it.

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpResponse.java (hunk 4 above)
Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpResponse.java]: Got it.

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java (hunk 5 above)
Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: Got it.

File: s3auth-relay/src/main/java/com/s3auth/relay/Main.java (hunk 6 above)
Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/Main.java]: Got it.

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java (hunk 7 above)
Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: Got it.

File: s3auth-relay/src/test/java/com/s3auth/relay/FtpClientTest.java (hunk 8 above)
Comment (pinaf) [inline s3auth-relay/src/test/java/com/s3auth/relay/FtpClientTest.java]: Removed class.

File: s3auth-relay/src/test/java/com/s3auth/relay/FtpClientTest.java (hunk 9 above)
Comment (pinaf) [inline s3auth-relay/src/test/java/com/s3auth/relay/FtpClientTest.java]: Removed class.

**Comment by pinaf** [comment]
@krzyk Thanks for the thorough review. I agree an the base class comment above and added a puzzel in `HttpFacade` for it.

File: s3auth-relay/pom.xml (hunk 2 above)
Comment (krzyk) [inline s3auth-relay/pom.xml]: Strange, I just used your branch, removed the version in commons-net and the project build fine.

File: s3auth-relay/src/test/java/com/s3auth/relay/FtpFacadeTest.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    @Test
+    @Ignore
+    public void connectDisconnect() throws IOException {
+        final FtpFacade facade = FtpFacadeTest.mockFacade();
+        final FTPClient ftp = new FTPClient();
+        try {
+            facade.listen();
+            ftp.connect("localhost");
+            final int reply = ftp.getReplyCode();
```
Comment (krzyk) [inline s3auth-relay/src/test/java/com/s3auth/relay/FtpFacadeTest.java]: please inline this variable

File: s3auth-relay/src/test/java/com/s3auth/relay/FtpFacadeTest.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            }
+        ).when(host)
+            .fetch(
+                Mockito.any(URI.class),
+                Mockito.any(Range.class),
+                Mockito.any(Version.class)
+            );
+        final Hosts hosts = Mockito.mock(Hosts.class);
+        Mockito.doReturn(host).when(hosts).find(Mockito.anyString());
+        final int port = PortMocker.reserve();
```
Comment (krzyk) [inline s3auth-relay/src/test/java/com/s3auth/relay/FtpFacadeTest.java]: and this one also can be inlined

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    /**
+     * Dispatch one request from the encapsulated queue.
+     * @return Amount of bytes sent to socket
+     * @throws InterruptedException If interrupted while waiting for the queue
+     * @checkstyle ExecutableStatementCount (100 lines)
+     * @checkstyle MultipleStringLiterals (100 lines)
+     */
+    @SuppressWarnings("PMD.AvoidCatchingThrowable")
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: We don't need this suppress

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+ * queue, processing it, and closing the socket. The class is instantiated
+ * by {@link FtpFacade} and is executed by Services Executor routinely.
+ *
+ * <p>The class is thread-safe.
+ *
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ * @see FtpFacade
+ * @todo 213:30min Implement authentication based on the USER and PASS
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: missing `#` in front of issue number

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+ *  commands.
+ * @todo 213:30min Implement relay functionality to fetch resource upon receival
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: missing `#` in front of issue number

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpResponse.java [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * <p>By default FTP code is OK (200) and content is empty.
+ *
+ * <p>The class is NOT thread-safe.
+ *
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ * @see FtpThread
+ * @todo 213:30min Implement methods 'withCode', 'withText', 'send' as above.
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpResponse.java]: missing `#` in front of issue number

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+        } catch (final InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    /**
+     * Report overflow problem to the socket and close it.
+     * @param socket The socket to report to
+     * @todo 213 implement
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: This one looks implemented, so todo could be removed, if not, then please add `#` and some more information how to implement it

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+ * <p>The class is instantiated in {@link Main}, once per application run.
+ *
+ * <p>The class is immutable and thread-safe.
+ *
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ * @see Main
+ * @checkstyle ClassDataAbstractionCoupling (10 lines)
+ * @todo 213:30min Implement TLS secure port listening in a manner analogous to
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: missing `#` in front of issue number

File: s3auth-relay/src/main/java/com/s3auth/relay/Main.java [hunk 19]
```diff
@@ -64,6 +64,8 @@ private Main() {
      * Entrance.
      * @param args Optional arguments
      * @throws Exception If something is wrong
+     * @todo 213:30min Create a FtpFacade in order to provide a FTP gateway.
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/Main.java]: missing `#` in front of issue number

**Comment by krzyk** [comment]
@pinaf some more comments from me

**Comment by pinaf** [comment]
@krzyk Thanks, I implemented all of your suggestions.
On the commons-net dependency, I tried to remove it again and both IntelliJ and `mvn clean install -Pqulice` fail to compile :(

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+ * <p>The class is thread-safe.
+ *
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ * @see FtpFacade
+ * @todo #213:30min Implement authentication based on the USER and PASS
+ *  commands.
+ * @todo #213:30min Implement relay functionality to fetch resource upon
+ * receival of the RETR command with valid authorization. Return an error
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: following lines of todo should be independent by additional space, just you did in other todos

**Comment by pinaf** [comment]
@krzyk fixed.

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by pinaf** [comment]
Travis OOM =P

File: s3auth-relay/pom.xml (hunk 2 above)
Comment (yegor256) [inline s3auth-relay/pom.xml]: so what about this? I think @krzyk is right, the version is not required here, it is inherited from parent POM: https://github.com/jcabi/jcabi-parent/blob/master/pom.xml#L1221-L1229

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+     * @param hosts Hosts
+     * @param port Port number
+     * @throws IOException If can't initialize
+     */
+    FtpFacade(@NotNull final Hosts hosts, final int port)
+        throws IOException {
+        this.server = new ServerSocket(port);
+        final FtpThread thread = new FtpThread(this.sockets, hosts);
+        final Runnable runnable = new VerboseRunnable(
+            new FtpFacade.FTPThreadRunnable(thread), true, true
```
Comment (yegor256) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: why not create an embedded class instead? would be much less verbose. now it's confusing and the name of the class is rather weird

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+    @SuppressWarnings({
+        "PMD.SingularField",
+        "PMD.UnusedPrivateField"
+    })
+    private final transient BlockingQueue<Socket> sockets;
+
+    /**
+     * Hosts to work with.
+     */
+    @SuppressWarnings({
```
Comment (yegor256) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: you can get rid of these suppressions just by doing `assert this.hosts != null` inside `dispatch()` method

**Comment by yegor256** [comment]
@pinaf overall, looks cool, good branch! just a few style comments above from me

File: s3auth-relay/pom.xml (hunk 2 above)
Comment (pinaf) [inline s3auth-relay/pom.xml]: I get a compile error both in IntelliJ and in `mvn clean install -Pqulice`. I'll remove the dependency though.

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java (hunk 21 above)
Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: You meam `FTPThreadRunnable`? I followed the desing of `HttpFacade`, which contains an embedded class `HttpThreadRunnable`. I'm not sure I understand the change you are suggesting.

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java (hunk 22 above)
Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: ok. should I do the same for `sockets`?

Comment (yegor256) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: sure

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java (hunk 21 above)
Comment (yegor256) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: sorry, not embedded, I mean an anonymous class

Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: got it

Comment (pinaf) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java]: @yegor256 check this out:
```
[WARNING] FindBugs: L P SIC: The class com.s3auth.relay.FtpFacade$1 could be refactored into a named _static_ inner class  At FtpFacade.java:[lines 116-125]
```

**Comment by pinaf** [comment]
@yegor256 Travis failed when I removed `commons-net` dependency :(
```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/yegor256/s3auth/s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java:[51,34] package org.apache.commons.net.ftp does not exist
[ERROR] /home/travis/build/yegor256/s3auth/s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java:[51,34] package org.apache.commons.net.ftp does not exist
[ERROR] /home/travis/build/yegor256/s3auth/s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java:[184,27] cannot find symbol
  symbol:   variable FTPReply
  location: class com.s3auth.relay.FtpFacade
```

**Comment by yegor256** [comment]
@pinaf OK for the embedded class. run `mvn dependency:tree` and check what is the version of commons-net you're getting. Maybe you should upgrade the parent to 0.32.1 (latest version)

**Comment by pinaf** [comment]
@yegor256 no luck :( commons-net does not show up in the ouput of that commnand even after setting jcabi-parent version to 0.32.1 in the <parent> tag of pom.xml.

**Comment by pinaf** [comment]
@yegor256 the fact that Travis also failed to build tells me there is something else going on here. I deleted my entire local maven repository and re-cloned the repo and it still can't find the `commons-net` class I need. I'm commiting again with the dependency.

**Comment by pinaf** [comment]
@yegor256 please take another look.

**Comment by yegor256** [comment]
@rultor merge pls

**Comment by davvd** [comment]
@krzyk Thank you, I've added **18 mins** to your account in payment/transaction "`50938104`". added +18 to your rating, now it is equal to [+3132](http://www.netbout.com/b/35035?open=rating)

**Comment by davvd** [comment]
@rultor deploy

**Comment by yegor256** [comment]
@rultor deploy again

**Comment by pinaf** [comment]
@yegor256 Failed to resolve a domain name for sonatype.org it seems.

**Comment by yegor256** [comment]
@rultor deploy

**Pull request merged**